src/firefly/java/edu/caltech/ipac/visualize/plot/projection/ProjectionUtil.java

```diff
@@ -145,4 +150,384 @@ public static boolean checkPlate(ImageHeader H1, ImageHeader H2) {
+            if (theta < -90.0) theta = -180.0 - theta;
+        } else {
+            z = sinlat * euler[3] + coslat * euler[4] * coslng;
+            if (Math.abs(z) > 0.99) {
+                /* Use an alternative formula for greater numerical accuracy. */
+                theta = acosd(Math.sqrt(x * x + y * y));
+                if (z < 0.0)
+                    theta = -Math.abs(theta);
+                else
+                    theta = Math.abs(theta);
+            } else {
+                theta = asind(z);
+            }
+        }
+
+        result[0] = phi;
+        result[1] = theta;
+        return result;
+    }
+
+    /**
+     * Reverse spherical transformation: native coordinates to celestial coordinates.
+     * Based on WCSLIB's sphrev function.
+     *
+     * @param phi native longitude in degrees
+     * @param theta native latitude in degrees
+     * @param euler Euler angles from celset
+     * @return array of [lng, lat] in degrees
+     */
+    private static double[] sphrev(double phi, double theta, double[] euler) {
+        double tol = 1.0e-5;
+        double lng, lat;
+        double[] retval = new double[2];
+        double cosphi, costhe, dlng, dphi, sinphi, sinthe, x, y, z;
+
+        costhe = cosd(theta);
+        sinthe = sind(theta);
+
+        dphi = phi - euler[2];
+        cosphi = cosd(dphi);
+        sinphi = sind(dphi);
+
+        /* Compute celestial coordinates. */
+        x = sinthe * euler[4] - costhe * euler[3] * cosphi;
+        if (Math.abs(x) < tol) {
+            /* Rearrange formula to reduce roundoff errors. */
+            x = -cosd(theta + euler[1]) + costhe * euler[3] * (1.0 - cosphi);
+        }
+        y = -costhe * sinphi;
+        if (x != 0.0 || y != 0.0) {
+            dlng = atan2d(y, x);
+        } else {
+            /* Change of origin of longitude. */
+            dlng = dphi + 180.0;
+        }
+        lng = euler[0] + dlng;
+
+        /* Normalize the celestial longitude to [0, 360) */
+        lng = lng % 360.0;
+        if (lng < 0.0) lng += 360.0;
+        /* Handle floating-point edge cases */
+        final double epsilon = 1e-10;
+        if (lng >= 360.0 - epsilon) lng = 0.0;
+        if (Math.abs(lng) < epsilon) lng = 0.0;
+
+        if (dphi % 180.0 == 0.0) {
+            lat = theta + cosphi * euler[1];
+            if (lat > 90.0) lat = 180.0 - lat;
+            if (lat < -90.0) lat = -180.0 - lat;
+        } else {
+            z = sinthe * euler[3] + costhe * euler[4] * cosphi;
+            if (Math.abs(z) > 0.99) {
+                /* Use an alternative formula for greater numerical accuracy. */
+                lat = acosd(Math.sqrt(x * x + y * y));
+                if (z < 0.0)
+                    lat = -Math.abs(lat);
+                else
+                    lat = Math.abs(lat);
+            } else {
+                lat = asind(z);
+            }
+        }
+
+        retval[0] = lng;
+        retval[1] = lat;
+        return retval;
```

Comment:
Here and in the other utility functions, instead of returning an array of two or three values you could define and return a record. Records are a new java class syntax for immutable data, they are perfect for returns like this.

```java
public record LonLat(double lon, double lat) {}
```
then 
```java
return new LonLat(lng,lat);
```
not very c like but more java/javascript like.